### PR TITLE
fix(discover): Tokenize and parse escaped quotes correctly

### PR DESCRIPTION
**Pull request by Zylphrex**
When tokenizing the raw query, escaped quotes were not being checked and thus
it was being treated as the end of a token. Additionally, when parsing the
tokens, all leading and trailing quotes were being stripped, meaning if the
token ended with a quote, it was being removed.

### CODE REVIEW

**Review by wmak — APPROVED** [review]
LGTM, as discussed in slack, lets backlog a task so that the query on the UI is the same one we're sending to the backend.

File: src/sentry/static/sentry/app/utils/tokenizeSearch.tsx [hunk 1]
```diff
@@ -299,30 +299,29 @@ function splitSearchIntoTokens(query: string) {
...
   let token = '';
-  let endOfPrevWord = '';
-  let quoteType = '';
-  let quoteEnclosed = false;
+  let quoted = false;
 
-  queryChars.forEach((char, idx) => {
-    const nextChar = queryChars.length - 1 > idx ? queryChars[idx + 1] : null;
-    token += char;
+  for (let idx = 0; idx < queryChars.length; idx++) {
```
Comment (dashed) [inline src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:304]: @Zylphrex Do we need to update the corresponding backend code here https://github.com/getsentry/sentry/blob/ce898db90cf0a79ef3ec9991683bccdfaf8c693b/src/sentry/search/utils.py#L329-L369 ?

Comment (dashed) [reply src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:304]: afaik, they should be mirrored 1:1.

Comment (Zylphrex) [reply src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:304]: Let me double check, you're likely right.

Comment (dashed) [reply src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:304]: I remember @EvanPurkhiser writing the corresponding frontend code in https://github.com/getsentry/sentry/pull/15812

Comment (dashed) [reply src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:304]: @wmak there actually is a comment. It's in the parent function.

Comment (Zylphrex) [reply src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:304]: Will do, that would have helped to find this corresponding backend code.

File: src/sentry/static/sentry/app/utils/tokenizeSearch.tsx [hunk 2]
```diff
@@ -344,20 +351,12 @@ function isSpace(s: string) {
  */
 function formatTag(tag: string) {
   const idx = tag.indexOf(':');
-  const key = stripSurroundingQuotes(tag.slice(0, idx));
-  const value = stripSurroundingQuotes(tag.slice(idx + 1));
+  const key = tag.slice(0, idx).replace(/^"+|(?<!\\)"+$/g, '');
+  const value = tag.slice(idx + 1).replace(/^"+|(?<!\\)"+$/g, '');
```
Comment (wmak) [inline src/sentry/static/sentry/app/utils/tokenizeSearch.tsx:357]: From slack, negative lookbehind isn't supported by safari 🙄 will have to look at another solution
https://caniuse.com/js-regexp-lookbehind

File: src/sentry/search/utils.py [hunk 3]
```diff
@@ -430,6 +434,10 @@ def split_query_into_tokens(query):
                 quote_enclosed = not quote_enclosed
                 if quote_enclosed:
                     quote_type = char
+        if quote_enclosed and char == "\\" and next_char == quote_type:
+            token += next_char
+            idx += 1
+        idx += 1
```
Comment (dashed) [inline src/sentry/search/utils.py:440]: Wondering if the corresponding test is updated as well.

**Review by wmak — APPROVED** [review]
Still LGTM with `removeSurroundingQuotes` +1

**Review by dashed — APPROVED** [review]

**Pull request merged**